url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPLinkManager.java

```diff
@@ -0,0 +1,2057 @@
+    private HashSet<OCPLinkStateHandler>        linkStateHandlers;
+    /** Mapping between task IDs and registered message handlers. */
+    private HashMap<Integer, OCPMessageHandler> tidHandlers;
+    /** The management task ID handler. */
+    private OCPMessageHandler managementTidHandler;
+    /** The watchdog to use. */
+    private GenericWatchdog watchdog;
+    /** Whether or not this link has been cleaned up. */
+    private volatile boolean cleanedUp = false;
+
+    /** The logger for this class. */
+    private static Logger log = Logger.getLogger("ocpLinkManager");
+    /** The statistics interface. */
+    private static StatsInterface stats = UtilitiesFactory
+            .getUtilInterface().getStatsInterface();
+
+    /**
+     * Creates an OCP Link Manager in the DISCONNECTED state using the specified
+     * properties object.
+     *
+     * @param properties
+     *            The properties object used.
+     * @param index
+     *            The index of this OCP link. This parameter is used to
+     *            distinguish between multiple OCP links defined in the same
+     *            properties object.
+     */
+    public OCPLinkManager(final Properties properties, final int index) {
+        this.prop = properties;
+
+        new Version();
+
+        linkIndex = index;
+        loadConfig(properties);
+
+        statPrefix = STAT_PREFIX + linkIndex;
+        stats.registerStat(statPrefix + STAT_BAD_MESSAGE, true);
+        stats.registerStat(statPrefix + STAT_OUTGOING_HEARTBEAT, true);
+        stats.registerStat(statPrefix + STAT_CONNECTION_SUCCESS, true);
+        stats.registerStat(statPrefix + STAT_CONNECTION_FAIL, true);
+        stats.registerStat(statPrefix + STAT_MASTER_SLAVE_SWAP, true);
+        stats.registerStat(statPrefix + STAT_QUEUE, false);
+
+        EnumSet<LegacyOCPMessageTypes> typeSet =
+                EnumSet.allOf(LegacyOCPMessageTypes.class);
+
+        for (LegacyOCPMessageTypes type : typeSet) {
+            if (type.getSupportRx()) {
+                stats.registerStat(statPrefix + STAT_MESSAGE_RX_BASE
+                        + type.getImplementation().getSimpleName(), true);
+            }
+            if (type.getSupportTx()) {
+                stats.registerStat(statPrefix + STAT_MESSAGE_TX_BASE
+                        + type.getImplementation().getSimpleName(), true);
+            }
+        }
+
+        linkState = LinkStates.DISCONNECTED;
+        linkStateHandlers = new HashSet<OCPLinkStateHandler>();
+        tidHandlers = new HashMap<Integer, OCPMessageHandler>();
+        managementTidHandler = null;
+        txMessages = new ArrayBlockingQueue<LegacyOCPMessage>(TX_QUEUE_LENGTH);
+    }
+
+    /**
+     * Loads the configuration.
+     * @param properties The properties object to use
+     * @return <code>true</code> if the configuration has changed
+     *          enough to require a link restart.
+     */
+    private boolean loadConfig(final Properties properties) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 454-558 into a separate method.